File: apps/server/api/src/routes/providers.ts
/**
 * Provider & Model Catalog Routes (/api/providers/*).
 * Controller delegating business logic to ProviderService.
 */
import { Hono } from "hono";
import { ProviderService } from "@/api/src/services/provider.service.js";
import type { ProviderId } from "@console/types";

export const providerRoutes = new Hono();
const providerService = new ProviderService();

/**
 * GET /api/providers — List all registered LLM providers.
 */
providerRoutes.get("/providers", (c) => {
  const providers = providerService.getProviders();
  return c.json({
    success: true,
    data: providers,
  });
});

/**
 * GET /api/providers/:id/models — Fetch dynamic models for provider.
 */
providerRoutes.get("/providers/:id/models", async (c) => {
  const providerId = c.req.param("id") as ProviderId;
  if (
    providerId !== "gemini" &&
    providerId !== "antigravity" &&
    providerId !== "opencode" &&
    providerId !== "codex"
  ) {
    return c.json({ success: false, error: `Invalid provider '${providerId}'.` }, 400);
  }

  try {
    const models = await providerService.getModels(providerId);
    return c.json({
      success: true,
      data: {
        provider: providerId,
        models,
      },
    });
  } catch (err) {
    const errorMsg = err instanceof Error ? err.message : String(err);
    return c.json({ success: false, error: errorMsg }, 500);
  }
});

File: apps/server-go/internal/routes/providers.go
// Provider & model catalog routes. Port of
// apps/server/api/src/routes/providers.ts: list providers and fetch
// dynamic models. Only implemented providers serve live data; the rest
// answer 501 until their Go ports land.
package routes

import (
	"github.com/gofiber/fiber/v2"

	"github.com/Adelodunpeter25/console/apps/server-go/internal/providers"
	"github.com/Adelodunpeter25/console/apps/server-go/internal/services"
)

func registerProviderRoutes(app *fiber.App, favorites *services.FavoriteService) {
	app.Get("/api/providers", func(c *fiber.Ctx) error {
		entries := providers.ListProviders()
		favs, err := favorites.List()
		if err == nil {
			for i, entry := range entries {
				entries[i].Models = providers.SortModelsByFavorites(entry.Models, favs)
			}
		}
		return c.JSON(fiber.Map{"success": true, "data": entries})
	})

	app.Get("/api/providers/:id/models", func(c *fiber.Ctx) error {
		id := c.Params("id")
		switch id {
		case "codex":
			models := providers.CodexModels(c.Context())
			if favs, err := favorites.List(); err == nil {
				models = providers.SortModelsByFavorites(models, favs)
			}
			return c.JSON(fiber.Map{"success": true, "data": fiber.Map{"provider": id, "models": models}})
		case "claude":
			models := providers.ClaudeModels(c.Context())
			if favs, err := favorites.List(); err == nil {
				models = providers.SortModelsByFavorites(models, favs)
			}
			return c.JSON(fiber.Map{"success": true, "data": fiber.Map{"provider": id, "models": models}})
		case "antigravity":
			return c.Status(fiber.StatusNotImplemented).JSON(fiber.Map{"success": false, "error": "Model discovery for '" + id + "' is not supported by the Go server yet."})
		default:
			return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{"success": false, "error": "Invalid provider '" + id + "'."})
		}
	})
}
